pycode/test/test_epidata_modifyDataframeSeries.py

```diff
@@ -18,72 +18,95 @@
-            moving_average=0, min_date='2021-01-04', max_date='2021-01-08',
+        # test impute zeros with moving average = 3
+        df = mDfS.impute_and_reduce_df(
+            self.test_df1, group_by_cols, mod_cols, impute='zeros',
+            moving_average=3, min_date='2021-01-05', max_date='2021-01-11',
             start_w_firstval=False)
-        pd_testing.assert_frame_equal(test_df_new, self.test_df3)
+        self.assertAlmostEqual(df[(df['Date'] == "2021-01-06") & (df['ID'] == 3.0)]['test_col3'].item(), 1 + 1 / 3)
+        self.assertAlmostEqual(df[(df['Date'] == "2021-01-06") & (df['ID'] == 3.0)]['test_col1'].item(), 2 + 1 / 3)
+        self.assertAlmostEqual(df[(df['Date'] == "2021-01-05") & (df['ID'] == 2.0)]['test_col3'].item(), 0)
+        self.assertAlmostEqual(df[(df['Date'] == "2021-01-05") & (df['ID'] == 2.0)]['test_col1'].item(), 0)
+        self.assertAlmostEqual(df[(df['Date'] == "2021-01-11") & (df['ID'] == 3.0)]['test_col3'].item(), 0)
+        self.assertAlmostEqual(df[(df['Date'] == "2021-01-11") & (df['ID'] == 3.0)]['test_col1'].item(), 0)
+        self.assertAlmostEqual(df[(df['Date'] == "2021-01-10") & (df['ID'] == 1.0)]['test_col1'].item(), 14)
+        self.assertAlmostEqual(df[(df['Date'] == "2021-01-10") & (df['ID'] == 1.0)]['test_col3'].item(), 2)
+
+        # test fill missing dates moving average = 2
+        # if moving average is an even number it always should calculate with one more earlier date
+        df = mDfS.impute_and_reduce_df(
+            self.test_df2, group_by_cols, mod_cols, impute='forward',
+            moving_average=4, min_date='2021-01-06', max_date='2021-01-13',
+            start_w_firstval=False)
+        self.assertAlmostEqual(df[(df['Date'] == "2021-01-08") & (df['ID'] == 3.0)]['test_col3'].item(), 2 + 3 / 4)
+        self.assertAlmostEqual(df[(df['Date'] == "2021-01-08") & (df['ID'] == 3.0)]['test_col1'].item(), 9 + 1 / 4)
+        self.assertAlmostEqual(df[(df['Date'] == "2021-01-06") & (df['ID'] == 2.0)]['test_col3'].item(), 0)
+        self.assertAlmostEqual(df[(df['Date'] == "2021-01-06") & (df['ID'] == 2.0)]['test_col1'].item(), 3)
+        self.assertAlmostEqual(df[(df['Date'] == "2021-01-13") & (df['ID'] == 3.0)]['test_col3'].item(), 1)
+        self.assertAlmostEqual(df[(df['Date'] == "2021-01-13") & (df['ID'] == 3.0)]['test_col1'].item(), 5)
+        self.assertAlmostEqual(df[(df['Date'] == "2021-01-10") & (df['ID'] == 1.0)]['test_col1'].item(), 12 + 1/4)
+        self.assertAlmostEqual(df[(df['Date'] == "2021-01-10") & (df['ID'] == 1.0)]['test_col3'].item(), 1 + 1/4)
+
+        # test start date higher than end date
+        # empty dataframe should be returned
+        df = mDfS.impute_and_reduce_df(
+            self.test_df1, group_by_cols, mod_cols, impute='forward',
+            moving_average=4, min_date='2021-01-13', max_date='2021-01-06',
+            start_w_firstval=False)
+        edf = pd.DataFrame()
+        self.assertEqual(len(edf),len(df))
+
```

Comment:
Please add two tests:

1.) start_w_firstval = true
2.) mod_cols = ['test_col1'] and check that test_col3 doesn't get changed.